src/iface/interface/mod.rs

```diff
@@ -429,11 +432,51 @@ impl Interface {
             }
         }
 
-        let mut readiness_may_have_changed = false;
+        // Limit the number of `poll()` loops to the largest packet ingress/egress size of any
+        // socket in the socket set to poll. This ensures that we will not indefinitely poll
+        // sockets.
+        let max_egress = sockets
+            .iter()
+            .map(|(_, socket)| match socket {
+                #[cfg(feature = "socket-raw")]
+                Socket::Raw(s) => s.packet_send_capacity(),
+                #[cfg(feature = "socket-icmp")]
+                Socket::Icmp(s) => s.packet_send_capacity(),
+                #[cfg(feature = "socket-udp")]
+                Socket::Udp(s) => s.packet_send_capacity(),
+                #[cfg(feature = "socket-tcp")]
+                Socket::Tcp(_) => 1,
+                #[cfg(feature = "socket-dhcpv4")]
+                Socket::Dhcpv4(_) => 1,
+                #[cfg(feature = "socket-dns")]
+                Socket::Dns(_) => 1,
+            })
+            .max()
+            .unwrap_or(1);
```

Comment:
Is this actually a reasonable way to calculate the number of polls? Or in other words, besides helping to prevent DoS, does this help the API consumers which care about completing the poll in any well-defined amount of time, for example to do some other real-time processing?

The function can't be redefined to do one iteration since this will be O(n^2) with a large amount of sockets, therefore it should probably have an additional variant that polls up to a predefined amount of times. It's not clear to me what the existing poll function should do, maybe call the new variant with `usize::MAX`?

Reply:
You make a good point here, and after discussion below we have changed it to poll a single loop of the socket set instead.